plutus-contract/test/Spec/TxConstraints/RequiredSigner.hs

```diff
@@ -0,0 +1,217 @@
+{-# LANGUAGE DataKinds           #-}
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TemplateHaskell     #-}
+{-# LANGUAGE TypeApplications    #-}
+{-# LANGUAGE TypeFamilies        #-}
+module Spec.TxConstraints.RequiredSigner(tests) where
+
+import Control.Monad (void)
+import Data.Map qualified as Map
+import Data.Void (Void)
+import Test.Tasty (TestTree, testGroup)
+
+import Data.Maybe (fromJust)
+import Data.String (fromString)
+import Ledger qualified
+import Ledger.Ada qualified as Ada
+import Ledger.CardanoWallet as CW
+import Ledger.Constraints.OffChain qualified as Constraints (paymentPubKey, plutusV1OtherScript,
+                                                             plutusV1TypedValidatorLookups, unspentOutputs)
+import Ledger.Constraints.OnChain.V1 qualified as Constraints
+import Ledger.Constraints.TxConstraints qualified as Constraints (mustBeSignedBy, mustIncludeDatum, mustPayToTheScript,
+                                                                  mustSpendScriptOutput, requiredSignatories)
+import Ledger.Tx qualified as Tx
+import Ledger.Typed.Scripts qualified as Scripts
+import Plutus.Contract as Con
+import Plutus.Contract.Test (assertFailedTransaction, assertValidatedTransactionCount, checkPredicateOptions,
+                             defaultCheckOptions, mockWalletPaymentPubKey, mockWalletPaymentPubKeyHash, w1, w2)
+import Plutus.Trace qualified as Trace
+import Plutus.V1.Ledger.Api (Redeemer (Redeemer), ScriptContext (scriptContextTxInfo), ToData (toBuiltinData), TxInfo)
+import Plutus.V1.Ledger.Contexts (txSignedBy)
+import Plutus.V1.Ledger.Scripts (ScriptError (EvaluationError), unitDatum)
+import PlutusTx qualified
+import PlutusTx.Prelude qualified as P
+import Prelude
+import Wallet.Emulator.Wallet (signPrivateKeys, walletToMockWallet)
+
+tests :: TestTree
+tests =
+    testGroup "required signer"
+        [
+        -- tests that use txSignedBy on-chain
```

Comment:
Why test the `txSignedBy` along `mustBeSignedBy` constraint? `mustBeSignedBy` uses `txSignedBy`.

Reply:
Initially thought there's be more different outcomes to check. Have changed to use `mustBeSignedBy` for most tests. One test exists to for `txSignedBy` phase-2 failure.

Reply:
I don't understand. Both validators are the exact same no? `mustBeSigned` uses the same implementation you used for `txSignedBy`.

Reply:
You're right. Now removed.